File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    N = len(genres)
    
    dict_ = dict()
    for i in range(N):
        if genres[i] in dict_:
            dict_[genres[i]].append((plays[i], i))
        else:
            dict_[genres[i]] = [(plays[i], i)]
    
    result = []
    for gen, li in dict_.items():
        li.sort(reverse=True)
        li.sort(key=lambda x:(-x[0], x[1]))
        # print(li)
        sum_ = 0
        for l in li: sum_ += l[0]
        
        tmp = []
        if len(li) < 2:
            tmp.append(sum_); tmp.append(li[0])
        else:
            tmp.append(sum_); tmp.append(li[0]); tmp.append(li[1])    
        result.append(tmp)
    
    result.sort(reverse=True)
    ans = []
    for r in result:
        for i in range(1, len(r)): ans.append(r[i][1])
    return ans

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>
#include <algorithm>

using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    int N = genres.size();
    
    unordered_map<string, vector<vector<int>>> Counter;
    unordered_map<string, int> Sum;
    
    for (int i = 0; i < N; i++)
    {
        Counter[genres[i]].push_back({plays[i], i});
        Sum[genres[i]] += plays[i];
    }
    
    vector<pair<string, int>> GenreOrder(Sum.begin(), Sum.end());
    sort(GenreOrder.begin(), GenreOrder.end(), [](pair<string, int> &a, pair<string, int> &b)
         {
             return a.second > b.second;
         });
    
    for (auto &pair : Counter)
    {
        sort(pair.second.begin(), pair.second.end(), [](vector<int> &a, vector<int> &b)
            {
                if (a[0] == b[0])
                {
                    return a[1] < b[1];
                }
                return a[0] > b[0];
            });
    }
    
    vector<int> answer;
    for (auto &pair : GenreOrder)
    {
        int cnt = 0;
        for (auto &vec: Counter[pair.first])
        {
            if (cnt < 2)
            {
                answer.push_back(vec[1]);
                ++cnt;
            }
            else
            {
                break;
            }
        }
    }
    
    return answer;
}